File: TradeMVVM.Trading/DataAnalysis/FindUnresolvedHoldings.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace TradeMVVM.Trading.DataAnalysis
{
    // Simple command-line utility to find holdings that appear in unresolved_isins.log
    // Usage:
    //  - Build as a small console app or run the compiled exe from the repository root.
    //  - Optional args: [holdingsCsv] [unresolvedLog] [outputPath]
    public static class FindUnresolvedHoldings
    {
        // Renamed from Main to Run to avoid defining multiple entry points in the solution.
        // Call FindUnresolvedHoldings.Run(...) from a small console project or from tests.
        public static int Run(string[] args)
        {
            try
            {
                string repoRoot = Directory.GetCurrentDirectory();

                string holdingsPath = args.Length > 0 && !string.IsNullOrWhiteSpace(args[0])
                    ? args[0]
                    : Path.Combine(repoRoot, "TradeMVVM.Trading", "DataAnalysis", "HoldingsReport.csv");

                string unresolvedLog = args.Length > 1 && !string.IsNullOrWhiteSpace(args[1])
                    ? args[1]
                    : Path.Combine(repoRoot, "TradeMVVM.Trading", "unresolved_isins.log");

                string outputPath = args.Length > 2 && !string.IsNullOrWhiteSpace(args[2])
                    ? args[2]
                    : Path.Combine(repoRoot, "TradeMVVM.Trading", "DataAnalysis", "unresolved_holdings_cs.txt");

                if (!File.Exists(holdingsPath))
                {
                    Console.Error.WriteLine($"Holdings file not found: {holdingsPath}");
                    return 2;
                }

                if (!File.Exists(unresolvedLog))
                {
                    Console.Error.WriteLine($"Unresolved log not found: {unresolvedLog}");
                    return 3;
                }

                var holdings = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                using (var sr = new StreamReader(holdingsPath))
                {
                    var header = sr.ReadLine();
                    if (header == null) return 0;
                    var headers = header.Split(';');
                    int isinIndex = -1;
                    for (int i = 0; i < headers.Length; i++)
                    {
                        if (string.Equals(headers[i].Trim(), "isin", StringComparison.OrdinalIgnoreCase) ||
                            string.Equals(headers[i].Trim(), "ISIN", StringComparison.OrdinalIgnoreCase))
                        {
                            isinIndex = i; break;
                        }
                    }

                    if (isinIndex < 0)
                    {
                        // fallback: assume first column is ISIN
                        isinIndex = 0;
                    }

                    string line;
                    while ((line = sr.ReadLine()) != null)
                    {
                        if (string.IsNullOrWhiteSpace(line)) continue;
                        var parts = line.Split(';');
                        if (parts.Length <= isinIndex) continue;
                        var isin = parts[isinIndex].Trim();
                        if (!string.IsNullOrEmpty(isin)) holdings.Add(isin);
                    }
                }

                var isinRegex = new Regex("ISIN:\\s*([A-Z0-9]+)", RegexOptions.IgnoreCase);
                var unresolved = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                foreach (var line in File.ReadLines(unresolvedLog))
                {
                    var m = isinRegex.Match(line);
                    if (m.Success)
                    {
                        var i = m.Groups[1].Value.Trim();
                        if (!string.IsNullOrEmpty(i)) unresolved.Add(i);
                    }
                }

                var bad = holdings.Where(h => unresolved.Contains(h)).OrderBy(x => x).ToList();

                using (var outw = new StreamWriter(outputPath, false))
                {
                    if (bad.Count == 0)
                    {
                        outw.WriteLine("No holdings appear in unresolved_isins.log");
                    }
                    else
                    {
                        foreach (var b in bad) outw.WriteLine(b);
                    }
                }

                Console.WriteLine($"Holdings: {holdings.Count} total");
                Console.WriteLine($"Unresolved ISINs in log: {unresolved.Count}");
                Console.WriteLine($"Holdings not scraped (count={bad.Count}):");
                foreach (var i in bad) Console.WriteLine(i);
                Console.WriteLine($"Wrote results to: {outputPath}");

                return 0;
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex}");
                return 1;
            }
        }
    }
}

File: TradeMVVM.Trading/DataAnalysis/find_unresolved_holdings.py
#!/usr/bin/env python3
"""
Find ISINs from HoldingsReport.csv that appear in unresolved_isins.log
Usage: run in repository root or from DataAnalysis folder. Outputs to DataAnalysis/unresolved_holdings.txt
"""
import os
import re
import csv

here = os.path.dirname(__file__)
holdings_path = os.path.join(here, 'HoldingsReport.csv')
unresolved_log = os.path.join(os.path.dirname(here), 'unresolved_isins.log')
output_path = os.path.join(here, 'unresolved_holdings.txt')

if not os.path.exists(holdings_path):
    print(f"Holdings report not found: {holdings_path}")
    print("Run generate_holdings.py first to produce HoldingsReport.csv")
    raise SystemExit(1)

if not os.path.exists(unresolved_log):
    print(f"Unresolved log not found: {unresolved_log}")
    raise SystemExit(1)

# read holdings ISINs
holdings = set()
with open(holdings_path, newline='', encoding='utf-8') as f:
    reader = csv.DictReader(f, delimiter=';')
    for row in reader:
        isin = (row.get('isin') or row.get('ISIN') or '').strip()
        if isin:
            holdings.add(isin)

# parse unresolved log for lines like 'ISIN: DE000...' or entries inside
isin_regex = re.compile(r"ISIN:\s*([A-Z0-9]+)", re.IGNORECASE)
unresolved = set()
with open(unresolved_log, encoding='utf-8', errors='ignore') as f:
    for line in f:
        m = isin_regex.search(line)
        if m:
            unresolved.add(m.group(1).strip())

# intersection: holdings that are unresolved
bad = sorted(h for h in holdings if h in unresolved)

with open(output_path, 'w', encoding='utf-8') as out:
    if not bad:
        out.write('No holdings appear in unresolved_isins.log\n')
    else:
        out.write('\n'.join(bad) + '\n')

print(f"Holdings: {len(holdings)} total")
print(f"Unresolved ISINs in log: {len(unresolved)}")
print(f"Holdings not scraped (count={len(bad)}):")
for i in bad:
    print(i)

print(f"Wrote results to: {output_path}")
